### PR TITLE
[WIP] Nested Record Field Copy and Update Expression

### PR DESCRIPTION
**Pull request by williamtetlow**
https://github.com/fsharp/fslang-suggestions/issues/379

TODOs
- [x] [Create RFC](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1049-nested-record-field--copy-and-update-expression.md)
- [x] Fix implementation to enable updating more than 1 nested field
- [x] Fix implementation to handle ambiguities between TypeName.FieldName & FieldName.FieldName
- [x] Resolution of access to nested field that is type with `[<RequireQualifiedAccess>]` attribute
- [x] Resolution of records accessed through namespace or module
- [x] Handle type abbreviations
- [x] Fix Build
- [x] Add Tests
- [ ] Add negative tests
- [x] Check all parts of nested field are declared in record
- [x] Check the same field is not declared twice in a nested field update
- [ ] Fix Intellisense suggestions for nested record field access
- [ ] Fix incorrect highlighting of TypeNames used in nested record field access
- [ ] Further testing of Intellisense
- [ ] Collaboration with anonymous records feature
- [ ] Investigate other language features that should support nested paths

### CODE REVIEW

**Comment by AviAvni** [comment]
This pr is working on first class lens I'm working with @williamtetlow as part of FSSF mentorship program

We already able to compile code like
```
type Street = { N: string }
type Address = { S: Street }
type Person = { A: Address; Age: int }
​
let person = { A = { S = { N = "Street 1" } }; Age = 30 }
​
let anotherPerson = { person with A = { person.A with S = { person.A.S with N = person.A.S.N + ", k.2" } } }
​
let anotherPerson1 = { person with A.S.N = person.A.S.N + ", k.2" }
```

The current situation "with" able to update 1 level of fields of record
There are 2 ways to specify the field
FieldName
TypeName.FieldName

We added the ability to update more then 1 level like
FieldName.FieldName
 
The way we did it is by transform the ast in the type checker see the example above

This is still work in progress
Some of the things we still need to work on
Fix the implementation to enable to update more then 1 nested field
Fix the build
Add tests
Handle the TypeName.FieldName vs FieldName.FieldName possible ambiguity
Check that the all parts of nested field is declared in record
Check that intellisence work as expected

We will be happy to get feedback

**Comment by forki** [comment]
cool stuff.

the current build error is already a good unit test ;-)

image

the fieldname is prefixed with a type

**Comment by forki** [comment]
something that also needs to be done (later when dust settles): for the standard case we have suggestions working. we should add that as well for nested records

**Comment by cartermp** [comment]
Modified the title to by WIP

@williamtetlow and @AviAvni absolutely love seeing this. Can you add a checklist of TODOs to the initial text of this PR so that we can easily see what's left to cover? Thanks!

**Comment by dsyme** [comment]
@cartermp We also need an RFC for this, and even a language suggestion.

I have some concerns, in particular about whether nested paths should be supported in other places, e.g. named argument setters:

    obj.Method(Prop1.NestedProp=3)
    SomeClass(Prop1.NestedProp=3, Prop1.NestedProp2=6)

It feels to me like if we're going to support one kind of nested update, we should support two.  And what about indexers?

    { mylist with Item(4).NestedProperty = 4 }
    obj.Method(Item(4).NestedProperty=3)

This seems useful and orthogonal in the context of both immutable and mutable data.  

Anyway, I'd like to discuss this properly from a design perspective, to make sure we are not just adding a feature that only applies to records when it can logically apply elsewhere too

**Comment by AviAvni** [comment]
@dsyme suggestion https://github.com/fsharp/fslang-suggestions/issues/379
RFC I'll work on it later

**Comment by AviAvni** [comment]
@dsyme  didn't understand to what
```
obj.Method(Prop1.NestedProp=3)
```
 will compile to

**Comment by smoothdeveloper** [comment]
@AviAvni 

```fsharp
let temp = obj.Method()
temp.Prop1.NestedProp <- 3
```

This is mostly used in constructors the same way C# has initializer syntax, but in F# it also works on plain methods, applying the setters on the return value.

**Comment by dsyme** [comment]
@AviAvni Currently 

   obj.Method(SettableProp1=x)

compiles to 

    let tmp = obj.Method()
    tmp.SettableProp1 <- x
    tmp

So 

   obj.Method(Prop1.SettableNestedProp1=x)

would be

    let tmp = obj.Method()
    tmp.Prop1.SettableNestedProp1 <- x
    tmp

Likewise there would a question about whether long paths could be used in named and/or optional arguments, e.g. would this have any meaning:

   member __.Method1(arg1 : SomeType)

   obj.Method(arg1.Prop1=x)

Though it's not clear what the meaning should be - it's like we would want functional update on a default argument but that doesn't fit that well with how F# does default arguments today.  I can see some use for this when option/argument spaces are very large (e.g. charting), e.g.

   Chart.Plot(Colors.Background=Color.Red, Colors.Foreground=Color.Blue, Layout.Packing=4)

The overall point is that we want to maintain  some degree of orthogonality (or at least the feeling of orthogonality) between record labels and named arguments...

**Comment by AviAvni** [comment]
@dsyme thanks we will work on it after finish with record things

**Comment by dsyme** [comment]
Note also we will need the same thing to work for anonymous records.. #4499.  I just did the copy-and-update  implementation for that today

**Comment by williamtetlow** [comment]
@dsyme How should we handle the case of nested access to a record field that is a type with the `[<RequiredQualifiedAccess>]` attribute?

**Comment by dsyme** [comment]
> @dsyme How should we handle the case of nested access to a record field that is a type with the [<RequiredQualifiedAccess>] attribute?

No.  The fact the field has type `SomeType` is enough.  `RequiredQualifiedAccess` is really "require either qualified access or a type-directed name resolution`

**Comment by AviAvni** [comment]
@dotnet-bot test please

**Comment by forki** [comment]
Wow build is green. Nice progress!

**Comment by williamtetlow** [comment]
### Progress Update

As well as the example posted by @AviAvni above the functionality has been extended to handle the following:

1. Resolution of RecordField access through Namespace, Module or TypeName
2. Resolution of FieldName when there are ambiguities between TypeName and FieldName
3. Expansion of AST for multiple field updates

````
module Test =
    module M =
        type AnotherNestedRecTy = { A : int; }

        type NestdRecTy = { B : string; C : AnotherNestedRecTy; }

        type RecTy = { D : NestdRecTy; E : string option; }


    let t1 = { M.RecTy.D = { M.B = "t1"; M.C = { M.A = 1; } }; M.E = None; }

    // Module.FieldName access with multiple updates
    let t2 = { t1 with M.D.B = "t2"; M.D.C.A = 2; }

    // Module.TypeName.FieldName access with multiple updates
    let t3 = { t1 with M.RecTy.E = Some "t3"; M.RecTy.D.B = "t3"; }
````

More examples can be found in the conformance tests.

The majority of Intellisense works. However, there are two issues found so far:

1. Identifier suggestions need to be implemented for when the access is more than one level deep
2. TypeName is not correctly highlighted when it's used in a nested field access

I am currently working on fixing the intellisense suggestions while adding in checks to ensure all fields are declared on the record type when expanding the AST.

The following tasks are still required to be completed:
- Check the same field is not declared twice in a nested field update
- Further testing of intellisense
- Add negative tests
- Collaboration with anonymous records feature
- Investigate other language features that should support nested updates

I've updated the TODOs to reflect this progress.

**Comment by cartermp** [comment]
@williamtetlow Fantastic progress so far, good work.

Could you also go through the [RFC](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1049-nested-record-field--copy-and-update-expression.md) and update it with the behavioral changes you've made in this PR? Moving forward, we want the RFCs to be a canonical source for _what_, _why_, and a high-level _how_ for all future changes to the language. This traceability is crucial for helping people understand behavior and help others on places such as StackOverflow.

Additionally, @zpodlovics has a good concern regarding struct records. Could this also be evaluated and written up in the RFC as something that is being accounted for?

Thank you!

**Comment by zpodlovics** [comment]
Right now the struct wrapper ctors are not optimized (https://github.com/dotnet/coreclr/issues/18542) and will require lot's of stack space and copy, especially with deep nesting: https://github.com/dotnet/coreclr/issues/18542#issuecomment-399535658

**Comment by williamtetlow** [comment]
RFC has been updated with behavioural changes and answer to @zpodlovics question about struct records. https://github.com/fsharp/fslang-design/pull/333

> The current implementation is expanding the simplified AST into the existing one and will have the same struct copy issue mentioned by @zpodlovics. However, as this issue is present in the original copy and update expression this would have to be fixed as well as updating the new implementation. This will be investigated further once the open tasks are completed

**Comment by realvictorprm** [comment]
Good job with this @williamtetlow 👍 After the whole SRTP topic is finished I want to help out 😃

**Comment by forki** [comment]
ping

**Comment by cartermp** [comment]
@williamtetlow @AviAvni Any progress on the remaining WIP items? We'll refrain from reviewing until you feel it's ready.

Now that Anonymous Records are merged and (soon) shipping, we'll want to make sure this can work with them too. I can update the RFC for that.

**Comment by ronaldschlenker** [comment]
Are record fields with union types somehow supported?

Like this:

```fsharp
type Response = {
    printHint: PrintHint option
}
and PrintHint = {
    requestPrintHint: RequestPrintHint option;
}
and RequestPrintHint = {
    printHeader: bool
}
```

Would it be possible to match cases in a given path of an update Expression and only update on match?

File: src/fsharp/NameResolution.fs [hunk 1]
```diff
@@ -3137,6 +3137,109 @@ let ResolveFieldPrim sink (ncenv:NameResolver) nenv ad ty (mp,id:Ident) allField
...
+                for KeyValue(_, v) in nenv.eFieldLabels do
+                    let fldOpt = v |> List.tryFind (fun r -> r.TyconRef.DisplayName = typeName)
+                    match fldOpt with
+                    | Some rfref -> yield rfref.RecdField.Id
+                    | None  -> () 
+            ]
+ 
+        let lookup() =
+            let frefs = 
+                try Map.find id.idText nenv.eFieldLabels |> success
```
Comment (vasily-kirichenko) [inline src/fsharp/NameResolution.fs]: Why not `Map.tryFind` with match?

**Comment by williamtetlow** [comment]
@cartermp there hasn't been any progress over the last few months, the intellisense issues are still outstanding and integration with anonymous records.

I was struggling to find a solution for the wrong suggestions being presented when you start going more than one field deep. e.g.

````fsharp
let anotherPerson1 = { person with A.S.
````

suggests `S` instead of `N`.

I will catch up with @AviAvni.

@vasily-kirichenko thanks for the feedback. I will take a look.

**Comment by realvictorprm** [comment]
Would be nice if we can get going here again. If I find time to do something again, I want to do something here too.

**Comment by KevinRansom** [comment]
@realvictorprm, it's been a long time since anyone looked at this?  do you want to take it over or should I close it?

Thanks

Kevin

**Comment by realvictorprm** [comment]
I forgot about that thingy. I think this can be considered as long hanging fruit with it being already quite reachable because of the existing progress. I can't promise anything but I'll try to find some time this year to look into it and try to update it and get it working even more.

**Comment by KevinRansom** [comment]
@realvictorprm, I am going to close this as an orphaned PR.  please make a branch in your fork if you want to revisit this PR.  We would welcome a PR, when you have something ready for us to work with.

Thanks everyone for your attention to this contribution.

Kevin

**Pull request closed without merging**